Reduce scope of Amount by avoiding unit conversion (#29)
tc39/proposal-measure: README.md

### Constructor

* `new Amount(value[, options])`. Constructs an Amount with the mathematical value of `value`, and optional `options`.

The object prototype would provide the following methods:

* `toString([ opts ])`: Returns a string representation of the measurement with any unit put in brackets (e.g., `"1.23[kg]`).
* `toLocaleString(locale[, options])`: Return a formatted string representation appropriate to the locale (e.g., `"1,23 kg"` in a locale that uses a comma as a fraction separator)
* `with(opts)`: Represent the same underlying mathematical value, possibly with different precision.

### Examples

Let's construct an Amount, query its properties, and render it.
First, we'll work with a bare number (no unit):

```js
let a = new Amount("123.456");
a.fractionDigits; // 3
a.significantDigits; // 6
a.with({ fractionDigits: 4 }).toString(); // "123.4560"
```

Notice that "upgrading" the precision of an Amount appends trailing zeroes to the number.

Here's an example with units:

```js
let a = new Amount("42.7", { unit: "kg" });
a.toString(); // "42.7[kg]"

> a.toString({ numberOnly: true }); // "42.7"

Author: I think this is something up for discussion. My vote would be to expose the unit in `toString`; I'd find it surprising if `toString` didn't expose such data.

For values without a unit, I'd be in favor of rendering just the underlying mathematical value, without brackets, and without a default "unit" unit à la `"123.456[1]"`;

> a.toString({ numberOnly: true }); // "42.7"

Author: What if we were to return a digit string for `.valueOf`?

> a.toString({ numberOnly: true }); // "42.7"

Reviewer: The default should be to include the full data model in the string.

For the name of the option, maybe explore something more like `unitDisplay: always|never|auto` or something.

> a.toString({ numberOnly: true }); // "42.7"

Author: I've added the option to the README. Let's discuss this in the JS numerics call.

**Author**: Ben Allen [@ben-allen](https://github.com/ben-allen)

## Goals and needs

Modeling measurements with a precision is useful for any task that involves measurements from the physical world. It can also be useful for other types of measurement; for example, (measurements of) currency amounts.

We propose to create a new object for representing measurements, for producing formatted string representations of measurements.

Common user needs that can be addressed by a robust API for measurements include, but are not limited to:

* The need to keep track of the precision of measured values. A measurement value represented with a large number of significant figures can imply that the measurements themselves are more precise than the apparatus used to take the measurement can support.

* The need to format measurements into string representations

## Description

We propose creating a new `Amount` API, whose values will be immutable and have the following properties:

Note: ⚠️  All property/method names up for bikeshedding.

> * `value` (String): The numerical value of the measurement, representing an exact mathematical value;

Author: I like that -- let's remove it.

* `unit` (String or undefined): The unit of measurement with which number should be understood (with *undefined* indicating "none supplied")
* `significantDigits` (Number): how many significant digits does this value contain? (Should be a positive integer)
* `fractionalDigits` (Number): how many digits are required to fully represent the part of the fractional part of the underlying mathematical value. (Should be a non-negative integer.)

#### Precision

A big question is how we should handle precision. When constructing an Amount, both the significant digits and fractional digits are recorded.

### Constructor

> * `new Measure(value)`. Constructs a Measure with the digit String `value` as the numerical value of the Measure.

Reviewer 2: The constructor options should also be explained here.

```suggestion
* `new Amount(value[, options])`. Constructs an Amount with the mathematical value of `value`, and optional `options`.
```

The object prototype would provide the following methods:

> * `toString()`: Returns a string representation of the measurement with any unit put in brackets (e.g., `"1.23[kg]`).

Author: Good point -- I've documented that now and added an example, too.

> * `toLocaleString()`: Return a string representation appropriate to the locale (e.g., `"1,23[kg]"` in locales that use a comma as a fraction separator)

Reviewer 2: This is the localized representation, so needs arguments.

```suggestion
* `toLocaleString(locale[, options])`: Return a formatted string representation appropriate to the locale (e.g., `"1,23 kg"` in a locale that uses a comma as a fraction separator)
```

* `with(opts)`: Represent the same underlying mathematical value, possibly with different precision.

### Examples

Let's construct a Measurement, query its properties, and render it. First, we'll work with a bare number (no unit):

```js
let a = new Amount("123.456");
a.fractionDigits; // 3
a.significantDigits; // 6
a.with({ fractionDigits: 4 }).toString(); // "123.4560"
```

> Notice that "upgrading" the precision of a Measurement essentially appends trailing zeroes to the number.

Reviewer 2: ```suggestion
Notice that "upgrading" the precision of an Amount appends trailing zeroes to the number.
```

#### Rounding

If one downgrades the precision of a Measurement, rounding will occur. (Upgrading just adds trailing zeroes.)

```js
let a = new Amount("123.456");
a.with({ significantDigits: 5 }).toString(); // "123.46"
```

By default, we use the round-ties-to-even rounding mode, which is used by IEEE 754 standard, and thus by Number and [Decimal](https://github.com/tc39/proposal-decimal). One can specify a rounding mode:

```js
let b = new Amount("123.456");
a.with({ significantDigits: 5, roundingMode: "truncate" }).toString(); // "123.45"
```

## Units

A core piece of functionality for the Measure proposal is to support units (`mile`, `kilogram`, etc.) as well as currency (`EUR`, `USD`, etc.). One can construct these

> ## Related but out-of-scope features

Author: Good idea -- done.

a.with({ significantDigits: 5, roundingMode: "truncate" }).toString(); // "123.45"
```

## Units (including currency)

A core piece of functionality for the proposal is to support units (`mile`, `kilogram`, etc.) as well as currency (`EUR`, `USD`, etc.). An Amount need not have a unit/currency, and if it does, it has one or the other (not both). Example:

```js
let a = new Amount("123.456", { unit: "kg" }); // 123.456 kilograms
let b = new Amount("42.55", { currency: "EUR" }); // 42.55 Euros
```

When serializing an Amount that has a unit/currency marker, units are automatically rendered in lowercase and currency always in uppercase.

> Note that, currently, no meaning is specified for units. You can use `"ABCDEFG"` as a currency and `"keelogramz"` as a unit.

Reviewer 2: We do check currency length, and localized formatting adds its own requirements on units.

```suggestion
Note that, currently, no meaning is specified for currencies or units.
You can use `"XYZ"` as a currency or `"keelogramz"` as a unit.
Calling `toLocaleString()` on an Amount with a unit not supported by Intl.NumberFormat will throw an Error.
```
